File: Algorithms/two-sum.cpp
1. Two Sum
https://leetcode.com/problems/two-sum/

Given an array of integers, return indices of the two numbers such that they add up to a specific target.

You may assume that each input would have exactly one solution, and you may not use the same element twice.

Example:
Given nums = [2, 7, 11, 15], target = 9,

Because nums[0] + nums[1] = 2 + 7 = 9,
return [0, 1].

#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        unordered_map<int, int> A;
        int sz = nums.size(), i, x, y;
        for (i = 0; i < sz; ++i) {
            x = nums[i];
            y = target - x;
            if (A.count(y)) {
                return {A[y], i};
            }
            A[x] = i;
        }
        return {};
    }
};

int main(void) {
    Solution solution;
    vector<int> nums, result;
    int target;

    nums = {2, 7, 11, 15};
    target = 9;
    result = solution.twoSum(nums, target);
    for (const auto & i : result) {
        cout << i << '\t';
    }
    
    return 0;
}

File: Algorithms/two-sum.cs
using System;
using System.Collections.Generic;

public class Solution {
    public static int[] TwoSum(int[] nums, int target) {
        Dictionary<int, int> A = new Dictionary<int, int>();
        for (int sz = nums.Length, i = 0; i < sz; ++i) {
            int val = nums[i], complement = target - val;
            if (A.ContainsKey(complement)) {
                return new int[] {A[complement], i};
            }
            A[val] = i;
        }
        return null;
    }
    public static void Main() {
        int[] nums, result;
        int target;
        
        nums = new int[] {2, 7, 11, 15};
        target = 9;
        result = TwoSum(nums, target);
        foreach (int i in result) {
            Console.Write(i);
            Console.Write(' ');
        }
        Console.WriteLine();
    }
}
